crates/in_stream/src/mem.rs

```diff
@@ -0,0 +1,348 @@
+        self.url.clone()
+    }
+
+    fn accept(&mut self) -> Result<<Self as InStreamListener>::Partial> {
+        loop {
+            // first, drain all pending connections from our recv channel
+            match self.recv.try_recv() {
+                Ok(stream) => {
+                    self.accept_queue.push(stream);
+                }
+                Err(crossbeam_channel::TryRecvError::Empty) => break,
+                Err(crossbeam_channel::TryRecvError::Disconnected) => {
+                    // wait until our user has accepted all pending connections
+                    // before letting them know the channel is broken
+                    if self.accept_queue.is_empty() {
+                        return Err(ErrorKind::BrokenPipe.into());
+                    }
+                }
+            }
+        }
+        if self.accept_queue.is_empty() {
+            // acceptor is non-blocking we have nothing to return
+            return Err(Error::with_would_block());
+        }
+        // pull the next item off the queue
+        Ok(InStreamPartialMem(Some(self.accept_queue.remove(0))))
+    }
+}
+
+/// a partly connected memory stream
+#[derive(Debug)]
+pub struct InStreamPartialMem(Option<InStreamMem>);
+
+/// memory stream specific connect config
+pub struct MemConnectConfig {}
+
+impl Default for MemConnectConfig {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+impl InStreamPartial for InStreamPartialMem {
+    type Stream = InStreamMem;
+    type ConnectConfig = MemConnectConfig;
+
+    /// we want a url like mem://
+    const URL_SCHEME: &'static str = "mem";
+
+    fn with_stream(stream: Self::Stream) -> Result<Self> {
+        Ok(Self(Some(stream)))
+    }
+
+    fn connect(url: &Url2, _config: Self::ConnectConfig) -> Result<Self> {
+        Ok(Self(Some(get_mem_manager().connect(url)?)))
+    }
+
+    fn process(&mut self) -> Result<Self::Stream> {
+        match self.0.take() {
+            None => Err(Error::new(ErrorKind::NotFound, "raw stream is None")),
+            Some(stream) => Ok(stream),
+        }
+    }
+}
+
+/// a singleton memory transport
+/// could be used for unit testing or for in-process ipc
+#[derive(Debug)]
+pub struct InStreamMem {
+    url: Url2,
+    send: crossbeam_channel::Sender<Vec<u8>>,
+    recv: crossbeam_channel::Receiver<Vec<u8>>,
+    recv_buf: Vec<u8>,
+}
+
+impl InStream for InStreamMem {}
+
+impl InStreamMem {
+    /// private constructor, you probably want `connect`
+    fn priv_new(
+        url: Url2,
+        send: crossbeam_channel::Sender<Vec<u8>>,
+        recv: crossbeam_channel::Receiver<Vec<u8>>,
+    ) -> Self {
+        Self {
+            url,
+            send,
+            recv,
+            recv_buf: Vec::new(),
+        }
+    }
+}
+
+impl Read for InStreamMem {
+    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
+        let mut disconnected = false;
+        loop {
+            // first, drain everything from our channel
```

Comment:
Is the _drain everything_ idiom performant enough here? I'm not suggesting it isn't just want to make sure it's thought about. I suppose it depends on the how often `read` is called and the size / number of messages to pull from the channel.

Reply:
added `for _ in 0.100` instead of the `loop`